### ISSUE TITLE
Kubelet restarts for unknown reasons causing e2e test failures.

### ISSUE THREAD

**Issue by gmarek** [kubernetes/kubernetes#11215]
Discussed originally in #10792, #10899.

We observe e2e test failures which were tracked down to Kubelet being restarted. Further investigation revealed that Kubelet was not getting `healthz` probes from Monit, so current theory is that restart is orchestrated by Monit which does not receive responses from Kubelet for some reason (Monit prober hanging? some network configuration issues? Problems with Kubelet HTTP server?). 

@wojtek-t @thockin @davidopp @quinton-hoole

**Comment by davidopp** [comment]
@dchen1107

**Comment by wojtek-t** [comment]
cc @saad-ali (who was looking into similar problems as part of #10792 - see e.g. https://github.com/GoogleCloudPlatform/kubernetes/issues/10792#issuecomment-119752028)

The next interesting run is:
kubernetes-e2e-gce-parallel/3419
with strange things happening on the master machine (all master components seems to be killed, Kubelet is up for ~6 more minutes, then it's restarted and everything else is then get back to health)

**Comment by gmarek** [comment]
For interested parties here is interesting part of the logs:
```
I0714 02:12:36.602524       5 gce.go:486] Target pool ac272028429cd11e5badb42010af0863 already deleted.$
^@^@^@^@^@^@^@^@^@^@^...^@^@^@I0714 02:19:34.012295       6 nodecontroller.go:114] Sending events to api server.$
I0714 02:19:34.013075       6 plugins.go:180] Loaded volume plugin "kubernetes.io/host-path"$
```

Where @^ is really ^@ and there's 3740 of them:
```
gmarek@breakwater:~/Downloads$ cat -A 146.148.76.155-22-kube-controller-manager.log | head -n 246 | tail -n 1 | hexdump
0000000 405e 405e 405e 405e 405e 405e 405e 405e
*
0001d30 405e 405e 405e 405e 3049 3137 2034 3230
```

As @mwielgus suggested this log is probably caused by some buffer overwrite. Only component that does not have this kind of log is Kubelet, which may indicate that some OOM happened on the machine, and Kubelet was unaffected because we adjusted its OOM score.

**Comment by wojtek-t** [comment]
cc @lavalamp who is on-call today for decision if we should try to debug it before 1.0

**Comment by lavalamp** [comment]
Actually @erictune swapped with me. But I think @dchen1107 is the one to decide about that.

**Comment by dchen1107** [comment]
Kubelet /healthz now does 3 checks: 1) docker version 2) hostname 3) syncLoop. The last one is added since v0.20 through #10004, which reports kubelet unhealthy when syncLoop takes more than 5 mins. This is a workaround for some docker call hangs without return.

I suspect the unknown restart is caused by long syncLoop here.

**Comment by dchen1107** [comment]
Of course, kubelet could do better job logging such failure, and also capture the reason of /healthz failure at monit log. But I don't think it is so critical to patch 1.0. Any disagreement?

Also for initial e2e test failure, I think we can improve the test with some retry logic.

**Comment by davidopp** [comment]
SGTM

**Comment by wojtek-t** [comment]
@dchen1107 - this definitely not the reason it all the cases I've seen, because I've seen Kubelet restarts after ~3 minutes (which means 5-minute deadline for syncLoop couldn't be the case here).
Also, as I wrote above, it's not that Kubelet is responding to /healthz with "not-ok" status - it just doesn't receive the /healthz requests in many cases.

**Comment by gmarek** [comment]
+1 to what @wojtek-t wrote. @dchen1107 please take a look at gce-parallel 3431 run kubelet 48.4 logs. 
You will notice that Kubelet does not log any GETs on healthz for first few minutes, but starts getting them after it's restarted ~2.5 minutes after it's started, around the line 193.

**Comment by wojtek-t** [comment]
I did a deeper debugging of that problem based on the run gce-parallel 3498 - it has exactly the same symptoms and have monit logs for it. What happens is that (all the logs comming from kubelet 104.197.34.96):

- kubelet from containerVM image is running
- we restart it after appropriately copying our binaries:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/saltbase/salt/kubelet/initd#L40
- kubelet starts at 06:26:57:

        I0715 06:26:57.568844    3120 manager.go:127] cAdvisor running in container: "/"
        I0715 06:26:57.569101    3120 fs.go:93] Filesystem partitions: map[/dev/disk/by-uuid/d1f816ed-d274-4db7-bc93-4984de845be6:{mountpoint:/ major:8 minor:1}]
        ...

- however due to some problems with Docker, it takes more than 30 seconds to finish the initialization:

        I0715 06:26:57.570074    3120 manager.go:163] Version: {KernelVersion:3.16.0-0.bpo.4-amd64 ContainerOsVersion:Debian GNU/Linux 7 (wheezy) DockerVersion:Unknown CadvisorVersion:0.15.1}
        ...
        E0715 06:26:59.774889    3120 manager.go:139] Failed to execute Info() call to the Docker client: dial unix /var/run/docker.sock: no such file or directory
        ...
        I0715 06:27:29.978116    3120 container_bridge.go:58] Recreated cbr0 and restarted docker
        ...
        I0715 06:27:30.081464    3120 server.go:623] Started kubelet
        ...
        I0715 06:27:30.104964    3120 server.go:82] Starting to listen read-only on 0.0.0.0:10255

- so the port on which Monit is checking the health of Kubelet was opened after ~35 seconds from the start

- Monit itself has default 30 second timeout for the start/stop operations:
 
        "In the case of a process check, Monit will wait up to 30 seconds for the start/stop action to finish
        before giving up and report an error. You can override this timeout using the TIMEOUT option."

- this results in Monit reporting failure in starting Kubelet:

         [UTC Jul 15 06:26:57] info     : 'kubelet' trying to restart
         [UTC Jul 15 06:26:57] info     : 'kubelet' stop: /etc/init.d/kubelet
         [UTC Jul 15 06:26:57] info     : 'kubelet' start: /etc/init.d/kubelet
         [UTC Jul 15 06:27:27] error    : 'kubelet' failed to start

- then no /healthz checks are even send to Kubelet and it is restarted after 2m30s:

         I0715 06:29:58.039918    6546 manager.go:127] cAdvisor running in container: "/system"
         ...
         I0715 06:29:59.326956    6546 server.go:623] Started kubelet

- the next start is smooth and everything works since that time


@dchen1107 @yujuhong @saad-ali @davidopp @thockin 

- I looked through few such failures in all cases it seemed that it took ~35 seconds for Kubelet to restart. I'm going to send out a PR to increase a Monit start/stop operation to 60 seconds to avoid such cases.

- There are two open questions which I completely don't understand:

 - why Kubelet was restarted after 2m30 seconds after its "start" operation failed (and not immediately after)?

  - why Monit is sending /healthz to Kubelet once per 2m30s not once per 2m (it's clear from the logs in all runs - both successful and failed, e.g.:

          I0715 00:18:59.842611     959 server.go:635] GET /healthz: (1.198812ms) 0 [[monit/5.4] 127.0.0.1:47516]
          ...
          I0715 00:21:30.061850     959 server.go:635] GET /healthz: (899.793µs) 0 [[monit/5.4] 127.0.0.1:47690]
          ...
          I0715 00:24:00.272673     959 server.go:635] GET /healthz: (1.224384ms) 0 [[monit/5.4] 127.0.0.1:47835]

**Comment by yujuhong** [comment]
> why Monit is sending /healthz to Kubelet once per 2m30s not once per 2m (it's clear from the logs in all runs - both successful and failed, e.g.:

From looking at monit.log in my own cluster, monit seems to be checking the services sequentially in every cycle. Among the services, kube-addon is bound to fail on the minions, so monit waits 30 seconds for it to start in every cycle, causing the cycle to be 2m+30s. If you have multiple failed-to-start services, the cycle could be longer.

Here is an example (with kube-addon start timeout set to 90s, and foo set to 30s):
```
[UTC Jul 15 15:25:48] error    : 'kube-addons' process is not running
[UTC Jul 15 15:25:48] info     : 'kube-addons' trying to restart
[UTC Jul 15 15:25:48] info     : 'kube-addons' start: /etc/init.d/kube-addons
[UTC Jul 15 15:25:48] error    : Error: Could not execute /etc/init.d/kube-addons
[UTC Jul 15 15:27:19] error    : 'kube-addons' failed to start
[UTC Jul 15 15:27:19] error    : 'foo' process is not running
[UTC Jul 15 15:27:19] info     : 'foo' trying to restart
[UTC Jul 15 15:27:19] info     : 'foo' start: /etc/init.d/foo
[UTC Jul 15 15:27:19] error    : Error: Could not execute /etc/init.d/foo
[UTC Jul 15 15:27:49] error    : 'foo' failed to start
------------ next cycle --------------
[UTC Jul 15 15:29:49] error    : 'kube-addons' process is not running
[UTC Jul 15 15:29:49] info     : 'kube-addons' trying to restart
[UTC Jul 15 15:29:49] info     : 'kube-addons' start: /etc/init.d/kube-addons
[UTC Jul 15 15:29:49] error    : Error: Could not execute /etc/init.d/kube-addons
[UTC Jul 15 15:31:19] error    : 'kube-addons' failed to start
[UTC Jul 15 15:31:19] error    : 'foo' process is not running
[UTC Jul 15 15:31:19] info     : 'foo' trying to restart
[UTC Jul 15 15:31:19] info     : 'foo' start: /etc/init.d/foo
[UTC Jul 15 15:31:19] error    : Error: Could not execute /etc/init.d/foo
[UTC Jul 15 15:31:50] error    : 'foo' failed to start
------------ next cycle --------------
[UTC Jul 15 15:33:50] error    : 'kube-addons' process is not running
[UTC Jul 15 15:33:50] info     : 'kube-addons' trying to restart
[UTC Jul 15 15:33:50] info     : 'kube-addons' start: /etc/init.d/kube-addons
[UTC Jul 15 15:33:50] error    : Error: Could not execute /etc/init.d/kube-addons
[UTC Jul 15 15:35:20] error    : 'kube-addons' failed to start
[UTC Jul 15 15:35:20] error    : 'foo' process is not running
[UTC Jul 15 15:35:20] info     : 'foo' trying to restart
[UTC Jul 15 15:35:20] info     : 'foo' start: /etc/init.d/foo
[UTC Jul 15 15:35:20] error    : Error: Could not execute /etc/init.d/foo
[UTC Jul 15 15:35:50] error    : 'foo' failed to start
```

The first timestamp in the a cycle is exactly two minutes after the last timestamp in the previous cycle, but the period becomes 2min + foo's 30s timeout + kube-addon's 90s timeout = 4 min

**Comment by yujuhong** [comment]
> why Kubelet was restarted after 2m30 seconds after its "start" operation failed (and not immediately after)?

If I understand correctly (from the logs), monit does not immediately restart a service after a failed attempt to start the process. It simply logs it and move on until the next cycle (e.g. the kube-addon attempts in the log). If monit keeps trying to restart the service until it's running, it'd be stuck at kube-addon forever.

**Comment by gmarek** [comment]
@yujuhong - good point. We haven't though about it when looking into logs. It sounds that we probably understand the problem completely, which is good, as it means that @wojtek-t PR should solve it completely (as long as we'll be able to start up Docker daemon in under a minute).

**Comment by dchen1107** [comment]
@gmarek I don't think @wojtek-t's PR solve the race issue at booting time we are talking here. Again, the issue is only occurring at boot time and like I mentioned at the initial issue: https://github.com/GoogleCloudPlatform/kubernetes/issues/10792#issuecomment-120137088 Such race is expected since we don't fully control boot sequence, and kubelet and docker can be restarted by various sources today. But upper layers including kubelet should handle such problems, and self recovers. My concern on extending 30s timeout for kubelet to 60s is that you might further delay other daemon's start time. For example, docker is done, monit is starting kubelet. It used to wait for 30s, and switch to docker, now it has to wait for 60s. Kubelet has a logic internally waitting for docker up at its own boot time too. 

My pr #11320 actually does tighten the regular monit health check cycles by avoiding extra 30s waiting on unnecessary kube-addons. But I am ok not patching it to v1.0 too.

**Comment by yujuhong** [comment]
@dchen1107 posted a new comment which gave more details about the cons of @wojtek-t's PR while I was editing my comment. 

> @yujuhong - good point. We haven't though about it when looking into logs. It sounds that we probably understand the problem completely, which is good, as it means that @wojtek-t PR should solve it completely (as long as we'll be able to start up Docker daemon in under a minute).

Hmm....I don't think increasing the start timeout would solve the problem.
  1. A lot of restarts are caused by race conditions from multiple sources trying to start kubelet: init.d, salt, monit. Unless we solve the fundamental issue, adjusting timeout wouldn't help much.



  2. Start timeout, in this particular case, seems irrelevant. When restarting kubelet, monit does *NOT* check /healthz, syncloop, etc. I believe it only checks the PID file (and perhaps whether the process is actually running). I verified this by delaying listening to /healthz in kubelet by > 1 minute, and monit still thought that kubelet started successfully. Therefore, the "failed to start" message that @wojtek-t saw was more likely due to other reason, and one possibility was some race condition.
  3.

**Comment by gmarek** [comment]
In Kubelet logs we saw 3-4 Kubelet starts: first one is kubelet from container image, probably started by init.d. After that salt kills running Kubelet and starts new Kubelet twice - once to check if it boots up at all (or test something else - I don't know), this Kubelet lives from 10ms to 100ms. Second one should live. The thing that we observe is that last Kubelet start too slowly, which makes Monit kill and recreate it.

I'm not certain what's the reason for 'failed to start' message, I need to compare it with a healthy Kubelet. The thing I do know is that we wait with creating healthz HTTP server until Docker is responsive and that this sometimes take more than 30s.

**Comment by yujuhong** [comment]
> I'm not certain what's the reason for 'failed to start' message, I need to compare it with a healthy Kubelet. The thing I do know is that we wait with creating healthz HTTP server until Docker is responsive and that this sometimes take more than 30s.

My point is (of course I could be wrong) that monit does not rely on the 30s timeout for checking /healthz, so increasing the timeout doesn't help.
  - salt starts kubelet
  - monit does routine health check, cannot reach /healthz
  - monit starts kubelet
     - If kubelet process is running, monit moves on
     - Otherwise, it waits for at maximum 30s, and then moves on.
  - monit ides for 2 minutes until the next check
  - monit does routine health check, again.

Increasing the start timeout (30s) makes monit waits longer, but doesn't help at all.

**Comment by gmarek** [comment]
Well, there's a simple way to check it, as I don't think increasing timeout will do much harm:)

Another data point is that I'm not convinced that there's a 'race' per se (i.e. there's never two kubelets running in the same time). When looking at Kubelets logs I've never seen interleaving of PIDs that write logs, which suggests that everything that starts Kubelet kills the previous one. It is possible that one running when Monit checks for liveness is the wrong one - we'll need to check it tomor... well - later today.

**Comment by yujuhong** [comment]
> Another data point is that I'm not convinced that there's a 'race' per se

One race condition I pointed out was that monit kills and restarts kubelet prematurely, as another source has *just* started kubelet (hence no response from /healthz). I have logs to verify that.
Also, most of them tries to stop and then start kubelet. The race condition would most likely cause short-lived kubelet processes.

As for why increasing timeout may be bad, Dawn's https://github.com/GoogleCloudPlatform/kubernetes/issues/11215#issuecomment-121755065 listed some.

**Comment by wojtek-t** [comment]
I agree that what @dchen1107 and @yujuhong wrote above the increasing timeout is theoretically correct (i.e. theoretically increasing timeout might probably harm us).
But this is NOT what we are observing in practice - in fact I can point you to at least 10 test runs, when thanks to increasing timeout to 60 seconds we would avoid its restarts 2m30s after it was in fact started.
I'm not going to say that this fix is actually what should be our long term solution.
But I think that self-restarting Kubelet 3 minutes after its start (and I've seen a case when it restarted more than once - i.e. after 2m30s, after 5m, etc.) is a really bad user experience - it's not that its restarted because it doesn't work correctly. It's just restarted because its initialization process took 35s instead of few seconds.
So although I agree that my PR is not what we should do long-term, I think that since my PR would fix all of the cases I've seen - we should merge it to solve the existing issue short term.

**Comment by yujuhong** [comment]
As it stands, I am still not convinced that increasing start timeout could fix this problem, for the reasons stated in my previous comments :-) 
In particular, I don't understand the 30s vs 35s argument, as monit *does not* check /healthz when starting kubelet.

I am, however, willing to accept that increasing the start timeout might help in some cases because of its side-effects (e.g. longer period before next check) that we don't understand. Some other methods may have the same effect: increasing the monit period, increasing the check timeout, etc. Another possibility is that the system is so heavily-loaded that "starting" a kubelet process would need > 30s (which is another issue of its own).

@dchen1107 has the final say on the PR merge :-)

> in fact I can point you to at least 10 test runs, when thanks to increasing timeout to 60 seconds we would avoid its restarts 2m30s after it was in fact started.

@wojtek-t, feel free to ignore the questions below if you're busy. I am just curious about *why* this fix works, and whether my understanding of monit is completely wrong :-)
  - " 2m30s after it was in fact started." -> According to you, /healthz should be ready ~35s. Why wasn't /healthz ready after 2m30s? 
  - The start timeout only applies when monit already decides to restart kubelet (am I completely wrong?). Why would it help during the periodic health checking?
  - I'd love to see some logs where the fix worked, if it's convenient for you.

**Comment by gmarek** [comment]
I believe I can answer one of your questions:
* healthz was ready after 2m30s, but Monit (or anything else) wasn't interested in it. Please take a look at logs from the gce-parallel 3431 run.
I don't understand the second one, an I doubt the third one is possible: we only see such failures in Jenkins, and as such we'd need to merge it first (or hammer k8s-bot on a PR:)

**Comment by wojtek-t** [comment]
@yujuhong - trying to answer your questions below:

>
" 2m30s after it was in fact started." -> According to you, /healthz should be ready ~35s. Why wasn't /healthz ready after 2m30s?

As @gmarek wrote above from the Monit and Kubelet logs it seems that if Monit after 30s decided that kubelet wasn't successfully started, it doesn't even try to send /healthz and just simply restarts it. I agree this is a strange behavior, but this is what we observed (maybe this was fixed in newer versions of Monit?)

>
The start timeout only applies when monit already decides to restart kubelet (am I completely wrong?). Why would it help during the periodic health checking?

You are completely right - it will not help. My change only affects the startup. In fact we basically observe the problems only during startup. The real problem is that if the Kubelet is restarted after those 2m30s, some tests on Jenkins are already running. If it is a test that is actually contacting Kubelet (like the Proxy test), this is causing failures.

>
I'd love to see some logs where the fix worked, if it's convenient for you.

Sorry - unfortunately I didn't save them (in fact I didn't expect such a discussion about this 1-line change). But basically, Kubelet logs were as they are - "Kubelet started" log appeared after ~35s from start, Monit didn't have "'kubelet' failed to start" log at all, and thus it sent a /healthz after ?? (I don't remember if it was ~2m30s or sth else) instead of restarting the Kubelet.

**Comment by gmarek** [comment]
To make it clear - (at least some) tests are failing because `RunRC` allows only some small number of container failures and restarting Kubelet causes those failures.

**Comment by gmarek** [comment]
I think we should merge #11339 ASAP, to see if Kubelet failures can be a reason for Container failures. To my understanding it shouldn't and only Daemon that should be able to make containers fail is Docker itself, but we probably should test it.

**Comment by thockin** [comment]
Anything blocking that test?

**Comment by gmarek** [comment]
Lack of review (I'll do this) and it needs a rebase.

**Comment by quinton-hoole** [comment]
ping?

**Comment by dchen1107** [comment]
test pr #11339 should be rebased and merged.

**Comment by gmarek** [comment]
I believe we haven't seen this for quite some time. Can we close this? @wojtek-t @quinton-hoole

**Comment by wojtek-t** [comment]
I'm fine with closing.
